### PR TITLE
MaskingEncoder 구현

### PR DESCRIPTION
**Pull request by muk-jjang**
1. label_encode, onehot_encode, linear_encode를 추가하였습니다.

---
추후 작업에서 
1. encoder들을 env step에서 따로 동작하도록 구성하는게 맞을지?
2. network 안에서 동작하도록 구성하는게 맞을지?

의견도 달아주세요

### CODE REVIEW

File: src/utils/utils.py [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+        for idx, reason in enumerate(reasons_dict.keys(), start=1):
+            labels[reasons_dict[reason]] = idx
+            
+        return labels
+
+    @staticmethod
+    def onehot_encode(reasons_dict: dict) -> torch.Tensor:
+        """Encodes reasons into one-hot vectors"""
+        batch_size, num_loc = next(iter(reasons_dict.values())).shape
+        num_reasons = len(reasons_dict)
```
Comment (falconlee236) [inline src/utils/utils.py:32]: 이 변수는 사용하지 않는것 같은데 적어둔 이유가 있을까요?

File: src/utils/utils.py [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+        
+        # Stack the values of reasons_dict along the last dimension
+        onehot = torch.stack(list(reasons_dict.values()), dim=-1)
```
Comment (falconlee236) [inline src/utils/utils.py:35]: torch.stack에 빈 배열이 들어올 때 예외처리같은것도 해두면 좋을 것 같습니다
[torch.stack docs](https://pytorch.org/docs/stable/generated/torch.stack.html)

File: src/utils/utils.py [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+            
+        return onehot
+
+    @staticmethod
+    def linear_encode(reasons_dict: dict, linear_layer: nn.Linear) -> torch.Tensor:
+        """Encodes reasons using a linear layer"""
+        onehot = MaskingEncoder.onehot_encode(reasons_dict)
```
Comment (falconlee236) [inline src/utils/utils.py:42]: 기본적으로 class 안에서는 내부 메소드를 그냥 사용할 수 있는걸로 알고 있는데,
static method 어노테이션으로 선언된 메소드는 클래스 내부에서 사용할 수 없나요?
```suggestion
        onehot = onehot_encode(resons_dict)
```

**Review by falconlee236 — COMMENTED** [review]
몇가지 궁금한점 코멘트 남겼습니다!
다른 분들도 이렇게 궁금한거 있으면 코멘트 남겨주세요

File: src/utils/utils.py (hunk 1 above)
Comment (muk-jjang) [reply src/utils/utils.py:32]: 아 이전에 for문으로 구성했다가 torch.stack으로 바꾸는 과정에서 삭제를 깜빡한거 같습니다..!

수정하겠습니다~

File: src/utils/utils.py (hunk 2 above)
Comment (muk-jjang) [reply src/utils/utils.py:35]: 예엡 반영해놓겠습니다

File: src/utils/utils.py (hunk 3 above)
Comment (muk-jjang) [reply src/utils/utils.py:42]: 넵!

File: src/utils/utils.py [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+    def label_encode(reasons_dict: dict) -> torch.Tensor:
+        """Encodes reasons into integer labels"""
+        batch_size, num_loc = next(iter(reasons_dict.values())).shape
+        labels = torch.zeros((batch_size, num_loc), dtype=torch.long)
+        
+        # label encoding
+        for idx, reason in enumerate(reasons_dict.keys(), start=1):
+            labels[reasons_dict[reason]] = idx
+            
+        return labels
```
Comment (KSW-KSM) [inline src/utils/utils.py:26]: 각 원인별로 인코딩을 수행하는 방식은 의도대로이나,
그 constraint의 우선순위는 reasons_dict를 전달할때 이미 우선순위대로 정렬이 되어있어야하는건가요?
**lebel 인코딩으로 인해서 'constraint'의 거리관계를 학습하게될텐데 이 lebel부여 방식을 하이퍼파라메터로써 전달받는 느낌으로 이해하면될까요?**

**Review by KSW-KSM — APPROVED** [review]
lebel encoding부분은 제가 전달을 잘못한 것 같고, 나머지는 제가 생각한대로 잘 구현된 것 같습니다!

File: src/utils/utils.py (hunk 2 above)
Comment (falconlee236) [reply src/utils/utils.py:32]: 아직 예외처리 완료 안된거 맞죠??? @muk-jjang

Comment (muk-jjang) [reply src/utils/utils.py:32]: 넵 집가서 푸시하겠습니다

**Comment by camellia785** [comment]
추후 작업에서

encoder들을 env step에서 따로 동작하도록 구성하는게 맞을지?
network 안에서 동작하도록 구성하는게 맞을지?
의견도 달아주세요

-> 조금 찾아보니까,, 강화학습 특성상 env step에서 인코딩을 수행하면 모든 상태가 동일한 방식으로 처리되니까 학습 과정에서 오류가 조금 더 적지 않을까 하는 생각이 듭니다

**Comment by muk-jjang** [comment]
> 추후 작업에서
> 
> encoder들을 env step에서 따로 동작하도록 구성하는게 맞을지? network 안에서 동작하도록 구성하는게 맞을지? 의견도 달아주세요
> 
> -> 조금 찾아보니까,, 강화학습 특성상 env step에서 인코딩을 수행하면 모든 상태가 동일한 방식으로 처리되니까 학습 과정에서 오류가 조금 더 적지 않을까 하는 생각이 듭니다

모든 상태가 동일한 방식으로 처리에 대해서 조금더 자세히 설명해주실 수 있나요??!

**Comment by camellia785** [comment]
> > 추후 작업에서
> > encoder들을 env step에서 따로 동작하도록 구성하는게 맞을지? network 안에서 동작하도록 구성하는게 맞을지? 의견도 달아주세요
> > -> 조금 찾아보니까,, 강화학습 특성상 env step에서 인코딩을 수행하면 모든 상태가 동일한 방식으로 처리되니까 학습 과정에서 오류가 조금 더 적지 않을까 하는 생각이 듭니다
> 
> 모든 상태가 동일한 방식으로 처리에 대해서 조금더 자세히 설명해주실 수 있나요??!

논문을 찾아봤는데, 

[1] https://openreview.net/forum?id=i84V7i6KEMd (2020)(COLT)
- "encoding environment dynamics in the reward function (REED) dramatically reduces the number of preference labels required"
- RED기반 보상이 성능을 높혔다

[2] https://users.cs.duke.edu/~parr/NIPS2016_Encoder.pdf (2016)(NIPS)
- 선형 가치 함수 근사와의 호환에서, 특징 인코딩의 중요성이 강조된 논문. 
- env step에서 인코딩을 수행하면 이러한 선형 근사 방법과 더 잘 호환될 수 있음.
- 왜? 강화학습에서 그런 특징을 지닌다고 합니다.

File: src/utils/utils.py [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+3. linear encoder
+'''
+import torch
+import torch.nn as nn
+
+class MaskingEncoder:
+    """Encoder for masking reasons"""
+    
+    def __init__(self, hidden_size: int = 8):
+        self.linear = nn.Linear(num_reasons, hidden_size)
```
[the remark this answers is not in the document]
Comment (camellia785) [reply src/utils/utils.py:14]: Num_reasons, reasons_dict에 대한 명확한 정의가 무엇인가요? 
현재 제약조건의 조합에 따라서 들어간다고 보았을 때 조금 헷갈립니다.

지금 인코딩 별로 이 두 가지가 달라지는 것인지 아니면, 
아예 input단에서 전체적으로 조합을 기반으로 들어갈 수 있도록 해야하는지

그러니까, 3가지 인코딩 이전 단계에서 제약조건의 조합으로 들어갈 수 있도록 하는 과정이 필요하다는 생각이 들었습니다.

[the remark this answers is not in the document]
Comment (muk-jjang) [reply src/utils/utils.py:14]: > Num_reasons, reasons_dict에 대한 명확한 정의가 무엇인가요? 
> 
> 현재 제약조건의 조합에 따라서 들어간다고 보았을 때 조금 헷갈립니다.
> 
> 
> 
> 지금 인코딩 별로 이 두 가지가 달라지는 것인지 아니면, 
> 
> 아예 input단에서 전체적으로 조합을 기반으로 들어갈 수 있도록 해야하는지
> 
> 
> 
> 그러니까, 3가지 인코딩 이전 단계에서 제약조건의 조합으로 들어갈 수 있도록 하는 과정이 필요하다는 생각이 들었습니다.

reasons_dict는 masking된 이유와 그에 해당하는 텐서를 모아놓은 dictinary입니다

num_reasons는 reasons_dict의 key 수 즉 마스킹된 이유의 수입니다.

그러므로 인코딩 별로 인풋이 달라지지 않습니다.

get_action_mask에서 일괄적으로 마스킹을 한 td를 받아올 것이기 때문에

제약조건의 조합으로 들어간다는 것에 대해 조금더 자세히 설명해주실 수 있나요??

그리고 왜 그렇게 들어가야 한다고 생각했는지도 말씀 해주시면 좋을거 같습니다!!

[the remark this answers is not in the document]
Comment (camellia785) [reply src/utils/utils.py:14]: 제약조건의 조합으로 라벨인코딩을 주어야한다고 하셔서,
만약 제약조건이 5가지라면, 그에 따른 조합은 5!만큼의 가지수가 나오게 되는 것이니까
인코딩 이전에 5!로(조합으로) 세분화해야하는 것이 아닐까! 라는 생각에 다음과같이 이야기한 것입니다.

인풋이 어떤식으로 들어가는지(tensodict) 굉장히 자세하게 조금 더 공부하고 이야기해봐야할 것 같습니다.!!

**Review by camellia785 — APPROVED** [review]

**Review by falconlee236 — APPROVED** [review]
